code/iaas/service-discovery/api/src/main/java/io/cattle/platform/servicediscovery/api/service/impl/ServiceDiscoveryApiServiceImpl.java

```diff
@@ -374,6 +387,57 @@ private void populateLogConfig(Map<String, Object> cattleServiceData, Map<String
             }
         }
     }
+    
+    @SuppressWarnings("unchecked")
+    private void populateSecrets(Map<String, Object> cattleServiceData, Map<String, Object> composeServiceData, Map<String, Object> secretsData) {
+        Object secrets = cattleServiceData.get(ServiceConstants.FIELD_SECRETS);
+        if (secrets instanceof List<?>) {
+            // we need to support two cases here. Long syntax and short syntax. If everything is default and filename matches secret name, then we 
+            // only export short syntax.
+            if (!((List<?>) secrets).isEmpty()) {
+                List<Object> list = (List<Object>) secrets;
+                List<Object> secretEntries = new ArrayList<>();
+                for (Object secret : list) {
+                    if (secret instanceof Map) {
+                        Map<String, Object> secretOpts = (Map<String, Object>) secret;
+                        String secretId = secretOpts.get(SECRET_ID).toString();
+                        Secret secretObj = objectManager.loadResource(Secret.class, secretId);
+                        String secretName = secretObj.getName(); 
+                        if (isShortSyntax(secretOpts)) {
+                            secretEntries.add(secretName);
+                        } else {
+                            String uid = secretOpts.get(UID).toString();
+                            String gid = secretOpts.get(GID).toString();
+                            String mode = secretOpts.get(MODE).toString();
+                            String filename = secretOpts.get(NAME).toString();
+                            Map<String, String> secretMap = new HashMap<>();
+                            secretMap.put(SOURCE, secretName);
+                            secretMap.put(TARGET, filename);
+                            secretMap.put(UID, uid);
+                            secretMap.put(GID, gid);
+                            secretMap.put(MODE, mode);
+                            secretEntries.add(secretMap);
+                        }
+                        
+                        Map<String, Object> secretTemplatesOpts = new HashMap<>();
+                        //TODO: add file opts
```

Comment:
@alena1108 https://github.com/rancher/cattle/pull/2771#issuecomment-309598115 It is still relevant. Since the fileOpts is not yet in the api, we need several changes in cattle and compose to make it work. Not sure if we need to wait that to be done.